lib/srv/exec.go

```diff
@@ -49,83 +49,147 @@ const (
 // a goroutine to SSH request handler and back to the calling client
-type execResult struct {
-	command string
+type ExecResult struct {
+	// Command is the command that was executed.
+	Command string
 
-	// returned exec code
-	code int
+	// Code is return code that execution of the command resulted in.
+	Code int
 }
 
-type execReq struct {
-	Command string
+// Exec executes an "exec" request.
+type Exec interface {
+	// GetCommand returns the command to be executed.
+	GetCommand() string
+
+	// SetCommand sets the command to be executed.
+	SetCommand(string)
+
+	// Start will start the execution of the command.
+	Start(channel ssh.Channel) (*ExecResult, error)
+
+	// Wait will block while the command executes.
+	Wait() (*ExecResult, error)
 }
 
-// execResponse prepares the response to a 'exec' SSH request, i.e. executing
+// NewExecRequest creates a new local or remote Exec.
+func NewExecRequest(ctx *ServerContext, command string) Exec {
+	return &LocalExecRequest{
+		Ctx:     ctx,
+		Command: command,
+	}
+}
+
+// LocalExecRequest prepares the response to a 'exec' SSH request, i.e. executing
 // a command after making an SSH connection and delivering the result back.
-type execResponse struct {
-	cmdName string
-	cmd     *exec.Cmd
-	ctx     *ctx
+type LocalExecRequest struct {
+	// Command is the command that will be executed.
+	Command string
+
+	// Cmd holds an *exec.Cmd which will be used for local exectuion.
+	Cmd *exec.Cmd
+
+	// Ctx holds the *ServerContext.
+	Ctx *ServerContext
+}
+
+// GetCommand returns the command string.
+func (e *LocalExecRequest) GetCommand() string {
+	return e.Command
 }
 
-// parseExecRequest parses SSH exec request
-func parseExecRequest(req *ssh.Request, ctx *ctx) (*execResponse, error) {
-	var e execReq
-	if err := ssh.Unmarshal(req.Payload, &e); err != nil {
-		return nil, trace.BadParameter("failed to parse exec request, error: %v", err)
+// SetCommand gets the command string.
+func (e *LocalExecRequest) SetCommand(command string) {
+	e.Command = command
+}
+
+// start launches the given command returns (nil, nil) if successful. execResult is only used
+// to communicate an error while launching
+func (e *LocalExecRequest) Start(channel ssh.Channel) (*ExecResult, error) {
+	var err error
+
+	// parse the command to see if the user is trying to run scp
+	err = e.parseSecureCopy()
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
 
-	// split up command by space to grab the first word
-	args := strings.Split(e.Command, " ")
+	// transforms the Command string into *exec.Cmd
+	e.Cmd, err = prepareCommand(e.Ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-	if len(args) > 0 {
-		_, f := filepath.Split(args[0])
+	// hook up stdout/err the channel so the user can interact with the command
+	e.Cmd.Stderr = channel.Stderr()
+	e.Cmd.Stdout = channel
+	inputWriter, err := e.Cmd.StdinPipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	go func() {
```

Comment:
I would move the inputWriter.Close() defer here so we would always close it when return